axlearn/common/attention.py

```
@@ -1216,18 +1216,37 @@ class Config(BaseLayer.Config):
         dim: Required[int] = REQUIRED  # The dimensionality of the positional embedding.
         theta: float = 10000.0  # The scale of base frequency.
 
-    def forward(self, positions: Tensor) -> Tensor:
+    def default_query_positions(self, max_seq_len: int) -> Tensor:
+        """Compute default `positions` value to be inputed into forward when `positions` is
+        not provided to the corresponding QKVLinear class such as `RoFormerQKVLinear`
+        """
+        return jnp.arange(max_seq_len)[None]  # [batch_size=1, max_seq_len].
+
+    def forward(
+        self, positions: Optional[Tensor] = None, max_seq_len: Optional[int] = None
+    ) -> Tensor:
         """
         TODO(bwzhang): 1. verify the performance under float32.
 
         Args:
             positions: A tensor representing the token position IDs.
                 The shape is [batch_size, seq_len].
+            max_seq_len: Max length of sequence, required if positions is not provided.
 
         Returns:
             Rotary Positional Embedding. Shape is [seq_len, dim].
+
+        Raises:
+            ValueError: If positions is None and max_seq_len is None.
         """
         cfg = self.config
+        if positions is None:
+            if max_seq_len is None:
```

what happens if both `positions` and `max_seq_len` are provided? should we check that they are consistent?

---

In that case `positions` will take precedence and we ignore `max_seq_len`. We won't need `max_seq_len` if the client provides explicit positions. Will add that to docstring.

---

Maybe we can error if they  are both provided?

---

Sure can do this. @ruomingp any opinion on this?

---

It seems better to throw an error if they are both provided or when they are inconsistent. WDYT?

---

Done. @ruomingp Please take a look?